### ISSUE TITLE
Query Builder search project the distance of ancestor/descendant nodes from the reference node.  

### ISSUE THREAD

**Issue by fernandogargiulo1986** [aiidateam/aiida_core#336]
It would be useful when traversing the graph in the seek of ancestor/descendant nodes that one could print the length of the path linking the two nodes of each connected pair. This might be achieved by adding special projection keys such as 'ancestor_distance', 'descendant_distance'. 
Also, as all other projection keys, the user should be able to order the results according to the ancestor/descendant_distance. 

Here is an example:

```python
qb = QueryBuilder().append(Nodes, tag='ref_node')
qb.append(Node, tag='ancestors', ancestor_of='ref_node', project=['id', 'ancestor_distance'])
qb.order_by({'ancestors', {'ancestor_distance': 'asc'}})

qb.all()
```

**Comment by fernandogargiulo1986** [comment]
This enhancement was asked by an AiiDA user

**Comment by lekah** [comment]
That is already implemented, but apparently not well documented:
`
QueryBuilder().append(Node, tag='ancestor',project='id',filters={'id':12}).append(Node, descendant_of='ancestor', project='id',edge_tag='edge', edge_project='depth').order_by({'edge':'depth'}).all()
`
Let me know if this works, so I can close the issue.

**Comment by fernandogargiulo1986** [comment]
Ok it's there. I just could not find the a list of the values for edge_project. Do you agree on having it? (One cannot figure it out in a simple way I guess)

**Comment by lekah** [comment]
Having it where? 
Here's the thing: I never hardcoded all the possible projections, so there is no "list of values". If the ORM-instance has such a property, the QueryBuilder can show it, if not there's an exception. So it would be ideal if we could expose the documentation of the ORM classes to the user (and with ORM I mean real ORM, the ones in the backend).

**Comment by fernandogargiulo1986** [comment]
I know projections are not hard-coded which is good. However, for models (DbNodes, ... ) corresponding to the AiiDA API objects (Nodes, ... ), the method get_db_columns() exists. which returns the projections dynamically by using default methods of Django/SQLAlchemy.  This turned out to be needed for the REST API. One could easily extend this the other ORM classes as you suggested. Then we have to find a way to expose it to the user (long-standing discussion). Finally, specifically to this topic, what will happen when DbPath will be replaced by recursive SQL queries?

**Comment by lekah** [comment]
1) The method get _db_dbcolumns could solve the issue if implemented for all other ORM classes. I am not sure I understand 100% the question, but there is also the (documented) possibility to "expand" a query by using '**' (double star as a projection). This will instruct the QB instance to try and project every possible property of this entity. 

`
QueryBuilder().append(Node, tag='ancestor',project='id',filters={'id':12}).append(Node, descendant_of='ancestor', project='id',edge_tag='edge', edge_project='**', project='**').dict()
`

This above way is one possibility to get the key-value pairs of the related properties.

2) The replacement of DbPath with a recursive query will not change the API very much. With the new pull request #351 one can start experimenting. The keyword depth, which is the only one of importance, I think, gives the same results!

3) Improvements in the documentation of the QB will be merged with #351. Let's hope this is enough for now.

**Comment by fernandogargiulo1986** [comment]
Ok, this solves the issue in my opinion. For more specific requests on the documentation side we should open more in-topic issues. I close this
